### ISSUE TITLE
Integrate Keycloak with Traefik via ForwardAuth

### ISSUE THREAD

**Issue by blarghmatey** [mitodl/ol-infrastructure#1473]
# Description/Context
We are standardizing on Traefik as our proxy for all services. As part of that we need a way to add authentication and authorization to applications/services that don't include that capability (e.g. Dagster or Airbyte). Traefik supports a middleware called ForwardAuth that allows it to delegate to systems like KeyCloak for this purpose. We would like to have a reusable component that easily sets up ForwardAuth via Keycloak for any system being proxied through Traefik.

# Acceptance Criteria
- [x] Traefik is configured to use ForwardAuth with Keycloak as the authentication provider

# References
- https://doc.traefik.io/traefik/middlewares/http/forwardauth/
- https://stackoverflow.com/questions/75168350/traefik-forward-auth-middleware-for-openid-connect
- https://geek-cookbook.funkypenguin.co.nz/docker-swarm/traefik-forward-auth/

**Comment by shaidar** [comment]
As the issue describes, the idea is have a request be proxied through traefik, make its way to the forward-auth middleware and be sent to keycloak and then return to traefik and be forwarded to the app listening on the requested URL. 

#### Initial trial
Tried getting this to work with Airbyte. The complications I ran into were:
- The Airbyte docker images include an nginx component which made things fairly complex routing traffic and in production would not be needed. I did some digging to see about removing nginx but based on the limited research I did, it appeared that we might need to build our own docker image of airbyte for that to work.
- Our pulumi stack is tightly coupled with consul and vault and spinning up a dev stack for testing an airbyte deployment wasn't very straightforward. Additionally, given that we were testing with keycloak as well and haven't fully verified/secured the config, I didn't want to be using our Vault CI cluster as it contains usable secrets. The same applies to keycloak. The devs were experimenting with the keycloak stack and didn't want anything coupled with any of our vault clusters, so had to setup a keycloak EC2 instance manually with DNS and a valid cert.
- The docs (traefik) aren't very clear on using the forward-auth middleware. Initially I didn't fully understand that the middleware is yet another docker image that's listening on another port and has traffic forwarded to it. Additionally there are a few different forward-auth containers, but most of the online references mention this https://github.com/thomseddon/traefik-forward-auth however the last commit to that repo was two years ago which is a bit concerning.

After running into those few issues, decided to go back to the basics and test out a simple app to verify the flow and make sure we have the basics configured (keycloak and traefik forward-auth)

#### Simple application
A lot of the examples in the documentation mention using a `whoami` app to test the config out. I ran into some problems initially since when the docker containers start up, they reach out to an oidc endpoint on the keycloak instance and if that fails to resolve, the service doesn't start. The docker compose config I was using was mostly an example used in the forward-auth repo - https://github.com/thomseddon/traefik-forward-auth/blob/master/examples/traefik-v2/swarm/docker-compose-oidc.yml. I initially missed that in the logs, and took me a bit to realize that the issue was that I was using a self signed cert on the keycloak instance and it required a verified one which made me switch to using letsencrypt. With that resolved, the flow somewhat worked, however keycloak was throwing an invalid redirect_uri error. It appears that by default, forward-auth appends an `/_oauth` endpoint to the redirect_uri and that isn't a valid endpoint on the sample application and thus the error. The value can be overridden in the config and a tried replacing with a blank value to test it out, but that didn't resolve the problem. I noted that as an issue to be resolved later as the overall flow seems to be working.

#### Dagster
Started working on setting up Dagster as an application to use with this setup given that it's one of our apps that only has basic auth and we need something more involved to give devs access to the pipelines. Made the stupid mistake of installing dagster on the EC2 instance that's running traefik in docker and that made routing kinda complicated. Would imagine it's doable (need to learn more about docker networking) how a container would route traffic to a service listening on the host instead of within the docker network namespace. Think for this test, it might be easier/best to run dagster in docker especially since this is how we have our setup. Started working on getting Dagster to run in docker.

**Comment by shaidar** [comment]
I was able to get the following to work:
End user making a request to the Dagster UI -> Traefik -> Traefik Forward Auth -> Keycloak -> Dagster UI
The setup is Dagster running in docker based on [this config](https://github.com/AntonFriberg/dagster-project-example/blob/main/docker-compose.yaml) which I added to the traefik services in the docker-compose.yaml file. For this testing, I manually added a Keycloak user to the DevOps realm and the flow ended up working as expected. I initially ran into the same issue that I had with the basic `whoami` application and the resolution to that was in the client Keycloak config which required me to add a wildcard to the app endpoint URL since by default the `redirect_uri` appends a `_oauth` to the url. 

With that proof-of-concept working, am starting to work in writing the required Pulumi pieces to get this in place on QA and test it out. The one thing that we need to decide on is whether we're gonna use Keycloak as our IDP. We had this issue https://github.com/mitodl/ol-infrastructure/issues/1332 of having Github be the IDP, however we ran into a roadblock as far as limiting Github user access to specific members of an specified org.

**Comment by shaidar** [comment]
Had a chat yesterday to map out the next steps and the gist of it is that we will remove the Github integration whereby it's acting an IDP for now. I started working on the Pulumi code to do the following:
- Add traefik config to Dagster's `docker-compose` file
- Create an app in Keycloak for the traefik forward-auth config to use

**Comment by shaidar** [comment]
One of the issues I ran into was my misunderstanding of the use of the `http` provider. I initially tried using letsencrypt based on the example [here](https://doc.traefik.io/traefik/user-guides/docker-compose/acme-tls/#setup). However, that was throwing an untrusted cert error which I wasn't sure whether that was causing further issues routing requests to the middleware. So, figured I'd try using the odl wildcard cert instead just to at least remove the cert as a potential problem. Wrongfully assumed that I could use the `http` provider with the tls options, however after some tinkering and reading and re-reading, realized that the http provider is used for a different use case. It's mainly used, if the traefik container needs to communicate with an http/https endpoint and that's where the configuration can be specified. Having cleared up that issue, I tried my hand at using the file provider and setup the wildcard cert and key which ended up working as expected. However, that didn't resolve the issue. Looking through the logs and the traefik dashboard, seems like I have some missing/misconfigured directive where the tls route going from traefik isn't being routed or doesn't know how to go to its next hop, `forward-auth` middleware in this case. I'm also a bit unclear on what configs should go into the docker-compose file as opposed to the dynamic configs that live in the traefik config file. Working on clarifying those things and think that should lead to figuring out the missing config.

**Comment by shaidar** [comment]
Finally got TLS working and being terminated on the traefik endpoint with our odl wildcard cert and routing requests through the forward-auth middleware (no TLS) to the auth-provider (Keycloak - TLS) and going back to the configured `redirect_uri` after a successful auth. A few points of confusion that I ran into with this setup:
- traefik documentation mentioning that the static config is either passed through env vars, config file, or command line and those are mutually exclusive. So, I wasn't really clear on whether passing in a config file containing the TLS config through commands in the docker-compose yaml file would override my other command line configs that I was passing as part of the traefik container.
- didn't come across any command line options to pass in the tls cert and key
- was turning off docker `exposedbydefault` in the traefik container config and not enabling traefik explicitly in the forward-auth directive and thus wasn't seeing it listed as middleware in the traefik UI.

Overall, valuable however at times a bit frustrating lessons but glad I was able to figure it out before the weekend, as it would've been annoying having it unresolved going into the weekend.
